hedera-node/docs/proposals/block-stream-bucket-uploader.md

```diff
@@ -0,0 +1,398 @@
+The following requirements have been identified for this proposal:
+- We need a third BlockStreamWriterMode, which sends the blocks to one or more buckets.
+- The buckets must be configurable.
+- Both the writer mode and the addresses of the buckets must be configurable during runtime (i.e., the node must not be restarted to take the changes into effect).
+- Blocks will be written to a local file initially and uploaded once done.
+- We want to upload only one instance of a block. All consensus nodes will try to upload the same block to a bucket, and therefore only the first one can succeed.
+- If a consensus node notices the block is already available, it should check if the blocks are the same (via MD5 hash). If they are different, the block should be moved to a hash mismatch directory, and an alert must be triggered.
+- If uploading a block fails, it should remain in the original directory for retry, and an alert needs to be triggered.
+- If the upload was successful, the block file will be moved to an uploaded directory.
+
+### Dependencies
+
+This proposal uses the MinIO Java SDK to provide a unified interface for cloud storage operations, supporting both AWS S3 and GCP Cloud Storage:
+
+```xml
+<dependencies>
+    <dependency>
+        <groupId>io.minio</groupId>
+        <artifactId>minio</artifactId>
+        <version>8.5.9</version>
+    </dependency>
+</dependencies>
+```
+
+#### Rationale for Using MinIO
+
+- **Unified API**: MinIO provides a single S3-compatible API that works with multiple cloud storage providers, simplifying the integration process.
+- **Flexibility**: Supports both AWS S3 and GCP Cloud Storage, allowing for easy configuration and management of multiple cloud providers.
+- **Scalability**: MinIO's client is designed to handle large-scale storage operations efficiently, making it suitable for high-throughput environments.
+- **Community and Support**: MinIO is a widely used open-source project with a strong community and commercial support options.
+
+By leveraging the MinIO SDK, we can streamline the process of uploading block files to cloud storage, ensuring compatibility and performance across different providers.
+
+## Core Components
+
+### Directory Structure
+
+Current directory in which blocks are written:
+
+`@ConfigProperty(defaultValue = "/opt/hgcapp/blockStreams") @NodeProperty String blockFileDir`
+
+Directory structure:
+```
+<blockFileDir>
+├── ./             # Current blocks being processed/uploaded
+├── /uploaded/     # Successfully uploaded blocks
+└── /hashmismatch/ # Blocks with hash validation failures
+```
+
+### Extended BlockStreamWriterMode
+
+```java
+public enum BlockStreamWriterMode {
+    GRPC,   // Stream blocks via gRPC
+    FILE,   // Write blocks to local files
+    BUCKETS // Write blocks to cloud storage buckets
+}
+```
+
+### Configuration
+
+The configuration is split into two parts:
+1. Network properties that can be updated at runtime
+2. Secure credentials stored locally
```

Comment:
Could also add Node Properties now for the `File` and/or `Buckets` configuration option